urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/INDIBLOBValue.java

```diff
@@ -0,0 +1,167 @@
+  /**
+   * The encoded data
+   */
+  private String base64EncodedData;
+  
+  /**
+   * Constructs a new BLOB Value from its coresponding bytes and format.
+   * 
+   * @param BLOBData the data for the BLOB
+   * @param format the format of the data
+   */
+  public INDIBLOBValue(byte[] BLOBData, String format) {
+    this.format = format;
+    this.BLOBData = BLOBData;
+    this.base64EncodedData = null;
+  }
+
+  /**
+   * Constructs a new BLOB Value from a XML &lt;oneBLOB&gt; element.
+   * @param xml the &lt;oneBLOB&gt; XML element
+   * @throws IllegalArgumentException if the XML element is not correct.
+   */
+  public INDIBLOBValue(Element xml) throws IllegalArgumentException {
+    int size = 0;
+    String f;
+ 
+    try {
+      String s = xml.getAttribute("size").trim();
+      size = Integer.parseInt(s);
+    } catch (NumberFormatException e) {
+      throw new IllegalArgumentException("Size number not correct");
+    }
+
+    if (!xml.hasAttribute("format")) {
+      throw new IllegalArgumentException("No format attribute");
+    }
+
+    f = xml.getAttribute("format").trim();
+
+    byte[] val;
+
+    base64EncodedData = xml.getTextContent().trim();
+    
+    try {
+      val = Base64.decode(base64EncodedData);
+    } catch (IOException e) {
+      base64EncodedData = null;
+      throw new IllegalArgumentException("Not BASE64 coded data");
+    }
+
+    if (f.endsWith(".z")) { // gzipped. Decompress
+      Inflater decompresser = new Inflater();
+      decompresser.setInput(val);
+
+      byte[] newvalue = new byte[size];
+
+      try {
+        decompresser.inflate(newvalue);
+
+        val = newvalue;
+      } catch (DataFormatException e) {
+        throw new IllegalArgumentException("Not correctly GZIPped");
+      }
+
+      decompresser.end();
+
+      f = f.substring(0, f.length() - 2);
+    }
+
+    if (val.length != size) {
+      throw new IllegalArgumentException("Size of BLOB not correct");
+    }
+
+    format = f;
+    BLOBData = val;
+  }
+  /**
+   * Gets the BLOB data
+   * @return the BLOB data
+   */
+  public byte[] getBLOBData() {
+    return BLOBData;
+  }
+
+  /**
+   * Gets the BLOB data in base64
+   * @return the BLOB data
+   */
+  public String getBase64BLOBData() {
+    if (base64EncodedData == null) {
+      base64EncodedData = Base64.encodeBytes(getBLOBData());
+    }
+    
+    return base64EncodedData;
+  }
+  
+  /**
+   * Gets the BLOB data format
+   * @return the BLOB data format
+   */
+  public String getFormat() {
+    return format;
+  }
+  
+  /**
+   * Gets the size of the BLOB data
+   * @return the size of the BLOB data
+   */
+  public int getSize() {
+    return BLOBData.length; 
+  }
+  
+  /**
+   * Save the BLOB Data to a file.
+   * @param file The file to which to save the BLOB data.
+   * @throws IOException if there is some problem writting the file.
+   */
+  public void saveBLOBData(File file) throws IOException {
+    FileOutputStream fos = new FileOutputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at line 163. The resource is closed at line 165. There are other execution paths that do not contain closure statements, for example, when FileOutputStream.write throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).